### PR TITLE
feat(if_gen): external memory port and portmaps

### PR DESCRIPTION
**Pull request by P-Miranda**
- add support to generate external rom and ram sp ports and portmaps
- generate external ram 2p, 2p_be, 2p_tiled and t2p interface port and
  portmaps
- add support to generate external ram dp and tdp port and portmaps
- includes ram_dp, ram_dp_be, ram_dp_be_xil, ram_tdp, ram_tdp_be
  variants
- add parameter prefix in external memory port width
- use plain wire without width in external memory portmaps
- add param_prefix as an optional argument for if_gen.py

### CODE REVIEW

File: scripts/if_gen.py [hunk 1]
```diff
@@ -26,6 +27,34 @@
     "clk_rst_port",
     "clk_en_rst_portmap",
     "clk_rst_portmap",
+    "ext_rom_sp_port",
```
Comment (jjts) [inline scripts/if_gen.py:30]: please remove ext_

File: scripts/if_gen.py [hunk 2]
```diff
@@ -189,6 +218,503 @@
...
+        "name": "r_en",
+        "default": "0",
+        "description": "read enable",
+    },
+    {
+        "sp": 1,
+        "tdp": 0,
+        "dp": 0,
+        "signal": "input",
+        "width": "ADDR_W",
```
Comment (jjts) [inline scripts/if_gen.py:247]: width cant be fixed 

it must be PREFIX_ADDR_W

File: scripts/if_gen.py [hunk 3]
```diff
@@ -189,6 +218,503 @@
...
+        "name": "addr_a",
+        "default": "0",
+        "description": "address port A",
+    },
+    {
+        "sp": 0,
+        "tdp": 1,
+        "dp": 1,
+        "signal": "output",
+        "width": "DATA_W",
```
Comment (jjts) [inline scripts/if_gen.py:307]: can't be fixed, must be PREFIX_DATA_W

there can be many memory interfaces wit different widths

File: scripts/if_gen.py [hunk 4]
```diff
@@ -189,6 +218,503 @@
     },
 ]
 
+rom = [
+    {
```
Comment (jjts) [inline scripts/if_gen.py:222]: what is this list for ?

**Comment by jjts** [comment]
This needs to be discussed. I do not follow.

A first commit is never this exhaustive. One memory type will suffice to check if it fits the bill.

File: scripts/if_gen.py (hunk 4 above)
Comment (P-Miranda) [reply scripts/if_gen.py:222]: generate interfaces for external roms

File: scripts/if_gen.py (hunk 2 above)
Comment (P-Miranda) [reply scripts/if_gen.py:247]: - parameter prefix is added for ports
(check https://github.com/IObundle/iob-lib/pull/613/files#diff-ef2b7ad0452150172a95f8f3d807daa5c79c04fccb6539c29c968dfb31b6315fR1563)
- portmaps do not use `width` field

**Review by arturum1 — COMMENTED** [review]
I have not tried to run this version of `if_gen.py`, but the changes in the code seem good.
The added code is similar to the existing one for other interfaces, so it should work well.
I approve.

**Pull request closed without merging**